### PR TITLE
Proposal: Add Tag Indicating To Replace Union Fields in Patch

### PR DESCRIPTION
**Pull request by ymqytw**
Move from issue #229.

@kubernetes/sig-api-machinery-misc @kubernetes/sig-cli-misc

### CODE REVIEW

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+## Proposed Changes
+
+### APIs
+
+Add a go struct tag/openAPI value telling clients to replace the entire struct when doing a *PATCH*. For the *inline* fields, we have 2 options:
+
+1) Move the tag to the parent struct's. E.g. add a tag to `Volume` instead of `VolumeSource`. This will limit the flexibility to add more fields in the future.
+
+2) Add some logic when lookup the metadata in go struct to find the tags for inline fields.
```
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I'd like to avoid adding more paths that require access to the go struct, and cannot be expressed in the openapi schema

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+
+This approach could be used to fix the issues with `kubectl apply` simply by having apply always replace union fields instead of patching them.  E.g. instead of setting `spec.strategy.type` without touching the rest of `spec.strategy`, patches should always replace the entire `spec.strategy`.
+
+## Proposed Changes
+
+### APIs
+
+Add a go struct tag/openAPI value telling clients to replace the entire struct when doing a *PATCH*. For the *inline* fields, we have 2 options:
+
+1) Move the tag to the parent struct's. E.g. add a tag to `Volume` instead of `VolumeSource`. This will limit the flexibility to add more fields in the future.
```
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I don't see how you can compatibly add more non-union fields to a type you have defined as a union type, so I'm not sure that loss of flexibility is a dealbreaker

Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: can you summarize current usage of union types, along with any that inline peer to other fields (like this name+volumeSource example)? If a significant number fall into the name+union pattern, we could describe a "named union" that gives special significance to the "name" field, and could be understood via the openapi schema.

**Comment by liggitt** [comment]
cc @jwforres

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 1 above)
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: We will need to add the existing mergeStrategy for lists as extensions to the open-api schema.  Can we just do the same for this at the same time?

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 2 above)
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I think @ymqytw has a list of these.  I know rollout strategy for Deployment is another.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 1 above)
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: openapi schemas flatten inlined structs

Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I believe that is ok for the current use cases, we can just put it in the parent right?

Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: right... that's option 1 :)

Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Yup makes sense.  I miss interpreted your original comment.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 2 above)
Comment (ymqytw) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I have list in the appendix, but it doesn't have enough information you need.  I will update it according to your comment.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+### kubectl
+
+No required change.
+
+## Strategic Merge Patch
+
+Add logic to support replacing the entire union fields when there is a `replace` directive. The logic will be similar to the logic of merge strategy for lists. 
+Add lookup logic to support lookup metadata of inline fields.
+
+The change in Strategic Merge Patch package will affect the behavior of how kubectl creating a patch and how API server apply a patch.
```
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Add a section about backwards / forward compatibility behavior.  Are there any issues here?  I think we might need to be careful for Volumes where the name may not be present in the old patches that expect merge, and will be lost.  I wonder if we can solve this with a new patch strategy in some way.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 2 above)
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Would be useful to include in this document all of the union types as of 1.5 and what the tag additions would be for each of them.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+### Open API
+
+We would need to add extensions to the openapi spec we publish. This is something we already need to do for the `patchStrategy` and `mergeKey` struct tags.
+
+# Alternatives Considered
+
+# 1. Add Union Support to the API Server
+
+The limitation is that it will be impossible or very hard to make it work for inline union types.
```
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Is this true?  Can you explain this more?

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+Add logic to support replacing the entire union fields when there is a `replace` directive. The logic will be similar to the logic of merge strategy for lists. 
+Add lookup logic to support lookup metadata of inline fields.
+
+The change in Strategic Merge Patch package will affect the behavior of how kubectl creating a patch and how API server apply a patch.
+
+### Open API
+
+We would need to add extensions to the openapi spec we publish. This is something we already need to do for the `patchStrategy` and `mergeKey` struct tags.
+
+# Alternatives Considered
```
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: I'd mention that the proposals below are not mutually exclusive with the proposal above, and could be added at some point in the future.

**Comment by pwittrock** [comment]
@bgrant0607 Is this something you would want to review?  This seems like a minimally invasive way to get support for unions in apply and I don't think it prevents us from implementing Unions in the future.  

@lavalamp FYI, I already briefly talked with you about this, but in case you wanted the design doc as a follow up.  The only required api changes would be adding the struct tags and some changes to the patch merge logic.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 2 above)
Comment (thockin) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Can we maybe live with existing inlined unions and ban it for future use?

i.e., VolumeSource should not have been inlined, but should have been a nested object.

**Comment by bgrant0607** [comment]
@pwittrock Yes, thanks, I'd like to take a look.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 2 above)
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: @thockin I think that is the proposal here.  The compromise would be that each Volume would need to be entirely managed by apply + configuration file or not managed by it at all.  e.g. A Volume could not have some fields managed by apply and come fields managed by another writer (e.g. some controller).

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+}
+```
+
+### kubectl
+
+No required change.
+
+## Strategic Merge Patch
+
+Add logic to support replacing the entire union fields when there is a `replace` directive. The logic will be similar to the logic of merge strategy for lists.
```
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: How would this be implemented?  Does the server know to replace the entire object?  Or does the client explicitly clear these fields?

**Comment by ymqytw** [comment]
Updated the proposal to address the comments.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 7]
```diff
@@ -0,0 +1,415 @@
+Add Tag Indicating To Replace *Union* Fields in Patch (a.k.a Support Replacing Maps in Strategic Merge Patch)
+=============
+
+This approach could be used to fix the issues with `kubectl apply` simply by having apply always replace union fields instead of patching them.  E.g. instead of setting `spec.strategy.type` without touching the rest of `spec.strategy`, patches should always replace the entire `spec.strategy`.
+
+## Proposed Changes
+
+### APIs
+
+Add a go struct tag/openAPI value telling clients to replace the entire struct when doing a *PATCH*. For the *inlined* fields, we have 2 options:
```
Comment (thockin) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: If this is the proposal, should we encourage or discourage discriminator fields for oneof cases?  We have both, and there will be more.

Comment (bgrant0607) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: We should encourage discriminators:
https://github.com/kubernetes/kubernetes/issues/35345

**Comment by bgrant0607** [comment]
@ymqytw Sorry, this is pretty long. Could you please summarize the proposal?

**Comment by ymqytw** [comment]
@bgrant0607 Sorry about that. Will summarize it today.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 8]
```diff
@@ -1,78 +1,61 @@
...
-	...
-	// patchStrategyForListEntry:"replace" is an additional struct tag indicating we replace the entire entry in the list, if they have the same merge key. 
-	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge" patchMergeKey:"name" patchStrategyForListEntry:"replace" protobuf:"bytes,1,rep,name=volumes"`
-	...
-}
-type Volume struct {
-	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
-	VolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
-}
+We reuse `patchStrategy:"replace"` if possible. This kind of patch will be backward compatible.
```
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: would the replace strategy prevent merging of descendants?

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md [hunk 9]
```diff
@@ -1,78 +1,61 @@
...
+
+Otherwise, we introduce a new struct tag, `patchStrategyForListEntry:"replace"`, for a list of inlined unions. This kind of patch will not be backward compatible.
```
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: isn't the point of a union type to ensure only one of a set of fields is set *at that level* in the object? both replace tags seem to remove the possibility of merging further down in the tree if the specified member of the union type didn't change.

Comment (ymqytw) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Yes, this is potential problem. Thanks for pointing this out.
I guess an solution is:
1) replace the whole struct if the union field changes.
2) merge, if the union field is still the old one.

WDYT?

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 8 above)
Comment (ymqytw) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: Yes.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 9 above)
Comment (liggitt) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: that's what I would expect a union-type tag to accomplish...

**Comment by ymqytw** [comment]
Updated the proposal. PTAL. Thanks.

File: contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md (hunk 9 above)
Comment (pwittrock) [inline contributors/design-proposals/add-metadata-indicating-to-replace-union-in-patch.md]: @liggitt a union-type tag would be a super set of what is proposed here.  This is an incremental step in that direction, but requires fewer changes to the API.

This proposal allows us to defer some of the following design questions:

- Do we validate that only 1 field is set?
- How do we validate and express discriminators?  Do that tags define which field matches a discriminator value?
- What about patchKey fields?  Are these exempt from the union?
- What about fields that are not pure unions, and 2 fields may be set together where neither is a discriminator, but are mutually exclusive with other fields that may have defaulted values.

**Comment by pwittrock** [comment]
@bgrant0607 here is a quick summary:

This proposal is to define a tag where, if set, for a given type all keys/fields missing from the request will be cleared when patching the object.  The goal is to defer making decisions around validating the correctness of unions, while enabling more correct patch behavior for union types.

| Capability                                      | Supported By This Proposal | Supported By Full Union  |
|---------------------------------|-----------------------------|--------------------------|
| Auto clear missing fields on patch   |  X                                            | X                                        |
| Merge union fields on patch        |  X                                            | X                                        |
| Validate only 1 field set on type  |                                                | X                                        |
| Validate discriminator field matches 1-of field   |                         | X                 |
| Support non-union  patchKey     |  X                                            |  TBD                                  |
| Support arbitrary combinations of set fields    |  X                                            |                                   |

This proposal supports any type where all of the desired fields must be set on every patch request.  It is agnostic to details such as if the type is a union or something else.  This makes things like supporting patchKeys simpler.

File: contributors/design-proposals/support-union-in-api-server.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge" patchMergeKey:"name" union: "inlined/name" protobuf:"bytes,1,rep,name=volumes"`
+	...
+}
+```
+
+We don't make any changes on other inlined unions.
+
+### Server Changes
+
+- Validate that neither *PATCH* nor *UPDATE* set multiple values within a Union
```
Comment (AdoHe) [inline contributors/design-proposals/support-union-in-api-server.md]: Here *UPDATE* means `PUT` operation right?

Comment (ymqytw) [inline contributors/design-proposals/support-union-in-api-server.md]: Yes. I will update it to make it clear.

File: contributors/design-proposals/support-union-in-api-server.md [hunk 11]
```diff
@@ -0,0 +1,200 @@
+Add Tag Indicating To Replace *Union* Fields in Patch
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Unclear how this relates to Unions.  

> Add new patchStrategy to clear fields not present in the patch

File: contributors/design-proposals/support-union-in-api-server.md [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+=============
+
+Add a tag `patchStrategy:"replaceKey"`. For a given type that has this tag, all keys/fields missing 
+from the request will be cleared when patching the object. For a field presents in the request,
+it will be merged with the live config.
+
+This approach could be used to fix the issues with `kubectl apply` (e.g. [#34292](https://github.com/kubernetes/kubernetes/issues/34292))
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Go into more detail about this use case for addressing unions and kubectl apply.  Start what problems this solves with patch - use case + example.

Use cases:
- As a user patching a map, I want keys mutually exclusive with those that I am providing to automatically be cleared.
- As a user running `kubectl apply`, when I update a field in my configuration file, I want mutually exclusive fields never specified in my configuration to be cleared by apply

Examples:
- General Example: Keys in a Union are mutually exclusive.  Clear unspecified union values in a Union that contains a discriminator.
- Specific Example: When patching a Deployment `.spec.strategy`, clear `.spec.strategy.rollingUpdate` if it is not provided in the patch so that changing  `.spec.strategy.type` will not fail.
- General Example: Keys in a Union are mutually exclusive.  Clear unspecified union values in a Union that does not contain a discriminator.
- Specific Example: When patching a Pod `.spec.volume`, clear all volume fields except the one specified in the patch.

File: contributors/design-proposals/support-union-in-api-server.md [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+#### Examples
+
+There are some examples using option 1 below.
+
+Example of non-inlined non-discriminated union:
+```go
+type ContainerStatus struct {
+	...
+	// Add patchStrategy:"replaceKey"`
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: `replaceKeys`

File: contributors/design-proposals/support-union-in-api-server.md [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+```
+
+Example of inlined union with `patchMergeKey` only.
+This case is special, because `Volumes` already has a tag `patchStrategy:"merge"`.
+We change the tag to `patchStrategy:"merge|replaceKey"`
+```go
+type PodSpec struct {
+	...
+	// Add another value "replaceKey" to patchStrategy
+	Volumes []Volume `json:"volumes,omitempty" patchStrategy:"merge|replaceKey" patchMergeKey:"name" protobuf:"bytes,1,rep,name=volumes"`
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: I guess we automatically recognize the mergeKey, and don't clear it in this case?

File: contributors/design-proposals/support-union-in-api-server.md [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+}
+```
+
+Example of non-inlined discriminated union:
+The tag contains the name of the discriminator.
+```go
+type DeploymentSpec struct {
+	...
+	// Add patchStrategy:"replaceKey/<discriminator-name>"
+	Strategy DeploymentStrategy `json:"strategy,omitempty" protobuf:"bytes,4,opt,name=strategy" patchStrategy:"replaceKey/type"`
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: WDYT of making the discriminator a separate tag.  `discriminator:type`

File: contributors/design-proposals/support-union-in-api-server.md [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+No required change.
+
+### Open API
+
+Update OpenAPI schema.
+
+### Strategic Merge Patch
+
+We need to add additional logic to support:
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Specify where this logic lives.  Server?  Client?  Both?  Which libraries or components are impacted?  Which sigs?

File: contributors/design-proposals/support-union-in-api-server.md (hunk 14 above)
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: What is the result of setting merge|replaceKeys vs just replaceKeys?

File: contributors/design-proposals/support-union-in-api-server.md [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+- Generate the correct patch respecting the new tag, `patchStrategy:"replaceKey"`
+- Merge the union if the union fields are unchanged
+- Replace the union if the selected field has changed.
+
+
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: What happens with version skew - client server?  Will things work correctly?  What sort of patch will be generated by the client?  How will it be merged by the server?

File: contributors/design-proposals/support-union-in-api-server.md [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+### kubectl
+
+No required change.
+
+### Open API
+
+Update OpenAPI schema.
+
+### Strategic Merge Patch
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Plz clarify where this code is executed.  Is it in the client or the server?

File: contributors/design-proposals/support-union-in-api-server.md [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+We need to add additional logic to support:
+- Generate the correct patch respecting the new tag, `patchStrategy:"replaceKey"`
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: We will need to make sure the client always sends a patch that includes all of the keys that it wants to keep.  Is this a change in how the client creates patches?  I think the client will need to explicitly specify the mergeStrategy to keep backwards compatibility with clients that may not send all of the keys.

File: contributors/design-proposals/support-union-in-api-server.md [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+| Capability | Supported By This Proposal | Supported By Full Union |
+|---|---|---|---|
+| Auto clear missing fields on patch | X | X |
+| Merge union fields on patch | X | X |
+| Validate only 1 field set on type |  | X |
+| Validate discriminator field matches one-of field |  | X |
+| Support non-union patchKey | X | TBD |
+| Support arbitrary combinations of set fields | X |  |
+
+## Proposed Changes
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Clarify the behavior for each of the new tags and old tags.  e.g. mergeKeys must always be present to perform the merge, and will be preserved.

File: contributors/design-proposals/support-union-in-api-server.md [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+We are leaning more to option 1, because the limitation of option 2 is that the metadata associated with
+the inlined APIs will not be reflected in the OpenAPI schema.
+
+#### Examples
+
+There are some examples using option 1 below.
+
+Example of non-inlined non-discriminated union:
+```go
+type ContainerStatus struct {
```
Comment (pwittrock) [inline contributors/design-proposals/support-union-in-api-server.md]: Include for each of these examples the actual contents of the: live object + merge patch + result.  Also include examples with apply file configuration + live object + result.

File: contributors/design-proposals/support-union-in-api-server.md (hunk 15 above)
Comment (ymqytw) [inline contributors/design-proposals/support-union-in-api-server.md]: Yes, it need additional logic to support discriminator.

If we use a separate tag `discriminator:type`, do we need `patchStrategy:replaceKeys` at the same time?

File: contributors/design-proposals/support-union-in-api-server.md (hunk 14 above)
Comment (ymqytw) [inline contributors/design-proposals/support-union-in-api-server.md]: Yes. Patching a list will lookup the metadata first. It should not be cleared.

File: contributors/design-proposals/support-union-in-api-server.md (hunk 19 above)
Comment (ymqytw) [inline contributors/design-proposals/support-union-in-api-server.md]: >We will need to make sure the client always sends a patch that includes all of the keys that it wants to keep. Is this a change in how the client creates patches?

Yes

>I think the client will need to explicitly specify the mergeStrategy to keep backwards compatibility with clients that may not send all of the keys.

Make sense to me. If so, I think we will need both the struct tag and a new patch directive to tell the API server how to do

File: contributors/design-proposals/support-union-in-api-server.md (hunk 14 above)
Comment (ymqytw) [inline contributors/design-proposals/support-union-in-api-server.md]: If we want to use `replaceKeys` in this case, then it will has additional meaning for list of non-primitives, merging list+ union.
I'm leaning to use `merge|replaceKeys` to keep the meaning of each tag clear.

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+#### Tags
+
+old tags:
+
+1) `patchMergeKey`:
+It is the key to distinguish the entries in the list of non-primitive types. It must always be
+present to perform the merge on the list of non-primitive types, and will be preserved.
+
+2) `patchStrategy`:
+It indicates how to generate and merge a patch for lists. It could be `mrege` or `replace`. It is optional for lists.
```
Comment (pwittrock) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: type: mrege => merge

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+
+new tags:
+
+`patchStrategy: replaceKeys`:
+
+We introduce a new value `replaceKeys` for `patchStrategy`. It indicates the all the fields that
```
Comment (pwittrock) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: It indicates **that** all fields needing to be preserved must **be** present in the patch.

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+```
+
+2) Non-inlined discriminated union:
+
+Type definition:
+```go
+type DeploymentSpec struct {
+	...
+	// Add discriminator:"<discriminator-name>"
+	Strategy DeploymentStrategy `json:"strategy,omitempty" protobuf:"bytes,4,opt,name=strategy" discriminator:"type"`
```
Comment (pwittrock) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: Doesn't this just become replaceKeys also?

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+```yaml
+unionName:
+  discriminatorName: bar
+  barField:
+    barSubfield: val2
+```
+
+3) Inlined union with `patchMergeKey` only.
+This case is special, because `Volumes` already has a tag `patchStrategy:"merge"`.
+We change the tag to `patchStrategy:"merge|replaceKeys"`
```
Comment (pwittrock) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: Shouldn't this also be just `replaceKeys`?

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md (hunk 24 above)
Comment (ymqytw) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: Removed.
Sorry, forget to update this.

File: contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md (hunk 25 above)
Comment (ymqytw) [inline contributors/design-proposals/add-new-patchStrategy-to-clear-fields-not-present-in-patch.md]: `merge` means we should merge the list.
`replaceKeys` mean we should replace keys for each entry in the list,
